node/pkg/accountant/audit.go

```diff
@@ -0,0 +1,292 @@
+			}
+			return
+		}
+
+		for _, pe := range pendingTransfers {
+			item, exists := transferDetails[pe.msgId]
+			if !exists {
+				if !pe.submitPending {
+					auditErrors.Inc()
+					acct.logger.Error("acctaudit: query did not return status for transfer, this should not happen, resubmitting it", zap.String("msgId", pe.msgId))
+					acct.submitObservation(pe)
+				} else {
+					acct.logger.Debug("acctaudit: query did not return status for transfer we have not submitted yet, ignoring it", zap.String("msgId", pe.msgId))
+				}
+
+				continue
+			}
+
+			if item.Status.Committed != nil {
+				digest := hex.EncodeToString(item.Status.Committed.Digest)
+				if pe.digest == digest {
+					acct.logger.Info("acctaudit: audit determined that transfer has been committed, publishing it", zap.String("msgId", pe.msgId))
+					acct.handleCommittedTransfer(pe.msgId)
+				} else {
+					digestMismatches.Inc()
+					acct.logger.Error("acctaudit: audit detected a digest mismatch, dropping transfer", zap.String("msgId", pe.msgId), zap.String("ourDigest", pe.digest), zap.String("reportedDigest", digest))
+					acct.deletePendingTransfer(pe.msgId)
+				}
+			} else if item.Status.Pending != nil {
+				acct.logger.Debug("acctaudit: contract says transfer is still pending", zap.String("msgId", pe.msgId))
+			} else if !pe.submitPending {
+				auditErrors.Inc()
+				acct.logger.Error("acctaudit: contract does not know about pending transfer, resubmitting it", zap.String("msgId", pe.msgId))
+				acct.submitObservation(pe)
+			}
+		}
+	}
+
+	acct.logger.Debug("acctaudit: exiting performAudit")
+}
+
+// handleMissingObservation submits a reobservation request if appropriate.
+func (acct *Accountant) handleMissingObservation(mo MissingObservation) {
+	// It's possible we received this transfer after we built the temporary map. If so, we don't want to do a reobservation.
+	if acct.transferNowExists(mo) {
+		acct.logger.Debug("acctaudit: contract reported unknown observation as missing but it is now in our pending map, ignoring it", zap.Uint16("chainId", mo.ChainId), zap.String("txHash", hex.EncodeToString(mo.TxHash)))
+		return
+	}
+
+	acct.logger.Debug("acctaudit: contract reported unknown observation as missing, requesting reobservation", zap.Uint16("chainId", mo.ChainId), zap.String("txHash", hex.EncodeToString(mo.TxHash)))
+	msg := &gossipv1.ObservationRequest{ChainId: uint32(mo.ChainId), TxHash: mo.TxHash}
+	acct.obsvReqWriteC <- msg
+}
+
+// transferNowExists checks to see if a missed observation exists in the pending transfer map. It grabs the lock.
+func (acct *Accountant) transferNowExists(mo MissingObservation) bool {
+	acct.pendingTransfersLock.Lock()
+	defer acct.pendingTransfersLock.Unlock()
+
+	chanId := vaa.ChainID(mo.ChainId)
+	txHash := ethCommon.BytesToHash(mo.TxHash)
+	for _, pe := range acct.pendingTransfers {
+		if (pe.msg.EmitterChain == chanId) && (pe.msg.TxHash == txHash) {
+			return true
+		}
+	}
```

Comment:
I'm not sure about this.  We know the transfer wasn't in the pending map when we started the audit and it feels unlikely to have been added in the short time since the audit was started so we'll end up iterating over the entire map for every missing observation.

The other end of this channel already deduplicates reobservation requests so maybe we don't need to do this check here.  See https://github.com/wormhole-foundation/wormhole/blob/fac7e10eb4ba195adcfae74b78e7771229650adb/node/cmd/guardiand/reobserve.go#L22-L25

Reply:
Yeah, I guess it is a pretty small window. I removed the check.